lib/ansible/modules/cloud/amazon/aws_acm.py

```diff
@@ -0,0 +1,356 @@
+  - boto3
+author:
+  - Matthew Davis (@matt-telstra)
+extends_documentation_fragment:
+    - aws
+    - ec2
+'''
+
+EXAMPLES = '''
+
+- name: upload a self-signed certificate
+  aws_acm:
+    certificate: "{{ lookup('file', 'cert.pem' ) }}"
+    privateKey: "{{ lookup('file', 'key.pem' ) }}"
+    name_tag: my_cert # to be applied through an AWS tag as  "Name":"my_cert"
+    region: ap-southeast-2 # AWS region
+
+- name: create/update a certificate with a chain
+  aws_acm:
+    certificate: "{{ lookup('file', 'cert.pem' ) }}"
+    privateKey: "{{ lookup('file', 'key.pem' ) }}"
+    name_tag: my_cert
+    certificate_chain: "{{ lookup('file', 'chain.pem' ) }}"
+    state: present
+    region: ap-southeast-2
+  register: cert_create
+
+- name: print ARN of cert we just created
+  debug:
+    var: cert_create.certificate.arn
+
+- name: delete the cert we just created
+  aws_acm:
+    name_tag: my_cert
+    state: absent
+    region: ap-southeast-2
+
+- name: delete a certificate with a particular ARN
+  aws_acm:
+    certificate_arn: "arn:aws:acm:ap-southeast-2:123456789012:certificate/01234567-abcd-abcd-abcd-012345678901"
+    state: absent
+    region: ap-southeast-2
+
+- name: delete all certificates with a particular domain name
+  aws_acm:
+    domain_name: acm.ansible.com
+    state: absent
+    region: ap-southeast-2
+
+'''
+
+RETURN = '''
+certificate:
+  description: Information about the certificate which was uploaded
+  type: complex
+  returned: when C(state) is C(present)
+  contains:
+    arn:
+      description: The ARN of the certificate in ACM
+      type: str
+      returned: when C(state) is C(present)
+      sample: "arn:aws:acm:ap-southeast-2:123456789012:certificate/01234567-abcd-abcd-abcd-012345678901"
+    domain_name:
+      description: The domain name encoded within the public certificate
+      type: str
+      returned: when C(state) is C(present)
+      sample: acm.ansible.com
+arns:
+  description: A list of the ARNs of the certificates in ACM which were deleted
+  type: list
+  returned: when C(state) is C(absent)
+  sample:
+   - "arn:aws:acm:ap-southeast-2:123456789012:certificate/01234567-abcd-abcd-abcd-012345678901"
+'''
+
+
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import boto3_conn, ec2_argument_spec, get_aws_connection_info
+from ansible.module_utils.aws.acm import ACMServiceManager
+
+
+# For converting PEM bodies to something that can be easily compared
+# e.g. ignore case, strip trailing whitespace
+#
+# There's probably a standard crypto library to do this, I'm writing 
+# another function here to minimize dependencies.
+# The lack of rigour we get by not importing a proper crypto lib
+# should result in only errors where equal certs are considered different,
+# in which case the only impact is that changed=True when it should be changed=False.
+# The other kind of error (different certs being considered equal) should not happen.
+def standardize_pem(pem):
```

Comment:
Using `module_utils.crypto.get_fingerprint()` is overkill IMO. It's better to convert the PEM to DER (binary format without the ASCII armour) and then compare; this can already be handled by the Python standard library: https://docs.python.org/3/library/ssl.html#ssl.PEM_cert_to_DER_cert (Should probably also work for Python 2.6, definitely works for 2.7.)

BTW, all code handling PEM files assumes that there is only one PEM object in there, and will use the first one. So using it for a chain won't give the results you want. In this module, this function is also used for comparing chains, so something different is needed anyway.

Reply:
The long `failed_when` conditions are used when I have `with_items`. If I split it up into an `assert`, then it gets messy, and then I'll have even deeper `item.item.item.item`s.

How about I split it out into aassert` tasks, *and* use a `vars` block to define `prev_task`, `original_cert`, etc. Will that make things clearer?